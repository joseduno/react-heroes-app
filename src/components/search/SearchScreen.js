import React, { useMemo } from 'react'
import queryString from 'query-string'
import { useLocation } from 'react-router-dom'
import { useForm } from '../../hooks/useForm'
import { HeroCard } from '../heroes/HeroCard'
import { getHeroByName } from '../../selectors/getHeroesByName'

export const SearchScreen = ({history}) => {

    /* Este componente muestra como recibir informaciÃ³n desde la url con query params */

    // obtiene la informacion de la url y la parsea como objeto
    // si q es undefined, se le asigna el string vacio ''
    const {search} = useLocation('search')
    const { q = '' } = queryString.parse(search)

    const [{searchText}, handleChange] = useForm({searchText: q})
    
    // const heroesFiltered = getHeroByName(searchText)
    const heroesFiltered = useMemo(() => getHeroByName(q), [q])

    const handleSearch = e => {
        e.preventDefault()
        // Al ejecutar el formulario, se agrega un query params por url
        history.push(`?q=${searchText}`)
    }


    return (
        <div className='row'>
            <div className='col-5 animate__animated animate__fadeInFast'>
                <h4>Search Form</h4>
                <hr />
                <form onSubmit={handleSearch}>
                    <input
                        type='text'
                        className='form-control'
                        placeholder='Find your hero'
                        onChange={handleChange}
                        autoComplete='off'
                        name='searchText'
                        value={searchText}
                    />
                    <button
                        type='submit'
                        className='btn btn-block btn-outline-primary mt-3'
                    >
                        Search
                    </button>
                </form>

            </div>
            <div className='col-7 animate__animated animate__fadeInRight'>
                <h4>Results</h4>
                <hr />

                {
                    (q === '')
                    &&
                    <div className='alert alert-info'>
                        Search a hero
                    </div>
                }

                {
                    (q !== '' && heroesFiltered.length === 0)
                    &&
                    <div className='alert alert-warning'>
                        Hero <b>{q}</b> was not found
                    </div>
                }


                {
                    heroesFiltered.map(hero => (<HeroCard key={hero.id} {...hero} />))
                }
            </div>
        </div>
    )
}
