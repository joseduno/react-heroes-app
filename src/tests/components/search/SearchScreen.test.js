import { mount } from "enzyme"
import { MemoryRouter, Route } from "react-router-dom"
import { SearchScreen } from "../../../components/search/SearchScreen"



describe('Pruebas en SearchScreen', () => {
    
    test('debe coincidir con el snapshot con valores por defecto', () => {
        const wrapper = mount(
            // Se necesita usar tanto el memory como el route con el mismo path en 
            // initialEntries y path respectivamente para indicar que debe renderizar un 
            // componente a partir de la url en entornos de pruebas, en este caso '/search'
            <MemoryRouter initialEntries={['/search']}>
                <Route path='/search' component={SearchScreen} />
            </MemoryRouter>
        )
        expect(wrapper).toMatchSnapshot()
        expect(wrapper.find('.alert-info').text().trim()).toBe('Search a hero')
    })

    test('debe mostrar a batman y el input con el valor batman', () => {
        const wrapper = mount(
            /*
            Otra vision de memory y route:
            En el initialEntries (MemoryRouter) se suele indicar la ruta del componente 
            que queremos probar con sus query params si es el caso, mientras que en el 
            path (Route), solo va la ruta del componente que se renderiza en ella, sin los 
            query params en caso de que los tenga
            */
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route path='/search' component={SearchScreen} />
            </MemoryRouter>
        )
        expect(wrapper.find('input').prop('value')).toBe('batman')
        expect(wrapper.find('HeroCard').length).toBeGreaterThanOrEqual(1)

    })

    test('debe mostrar un error si no se encuentra el hero', () => {
        
        const wrapper = mount(
            /*
            Otra vision de memory y route:
            En el initialEntries (MemoryRouter) se suele indicar la ruta del componente 
            que queremos probar con sus query params si es el caso, mientras que en el 
            path (Route), solo va la ruta del componente que se renderiza en ella, sin los 
            query params en caso de que los tenga
            */
            <MemoryRouter initialEntries={['/search?q=1jk23gb4kj123h4gv']}>
                <Route path='/search' component={SearchScreen} />
            </MemoryRouter>
        )
        expect(wrapper.find('.alert-warning').exists()).toBe(true)
        expect(wrapper).toMatchSnapshot()
    })

    test('debe hacer el submit correctamente con batman en el input text', () => {

        const historyMock = {push: jest.fn()}
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=1jk23gb4kj123h4gv']}>
                <Route path='/search' component={
                    () => <SearchScreen history={historyMock} />
                } />
            </MemoryRouter>
        )
        wrapper.find('input').simulate('change', {target: {
            name: 'searchText',
            value: 'batman'
        }})
        wrapper.find('form').simulate('submit', {preventDefault(){}})
        expect(historyMock.push).toHaveBeenCalledWith('?q=batman')

    })
    
})
