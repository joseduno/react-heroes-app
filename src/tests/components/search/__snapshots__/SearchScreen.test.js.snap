// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en SearchScreen debe coincidir con el snapshot con valores por defecto 1`] = `
<div
  className="row"
>
  <div
    className="col-5 animate__animated animate__fadeInFast"
  >
    <h4>
      Search Form
    </h4>
    <hr />
    <form
      onSubmit={[Function]}
    >
      <input
        autoComplete="off"
        className="form-control"
        name="searchText"
        onChange={[Function]}
        placeholder="Find your hero"
        type="text"
        value=""
      />
      <button
        className="btn btn-block btn-outline-primary mt-3"
        type="submit"
      >
        Search
      </button>
    </form>
  </div>
  <div
    className="col-7 animate__animated animate__fadeInRight"
  >
    <h4>
      Results
    </h4>
    <hr />
    <div
      className="alert alert-info"
    >
      Search a hero
    </div>
  </div>
</div>
`;

exports[`Pruebas en SearchScreen debe mostrar un error si no se encuentra el hero 1`] = `
<div
  className="row"
>
  <div
    className="col-5 animate__animated animate__fadeInFast"
  >
    <h4>
      Search Form
    </h4>
    <hr />
    <form
      onSubmit={[Function]}
    >
      <input
        autoComplete="off"
        className="form-control"
        name="searchText"
        onChange={[Function]}
        placeholder="Find your hero"
        type="text"
        value="1jk23gb4kj123h4gv"
      />
      <button
        className="btn btn-block btn-outline-primary mt-3"
        type="submit"
      >
        Search
      </button>
    </form>
  </div>
  <div
    className="col-7 animate__animated animate__fadeInRight"
  >
    <h4>
      Results
    </h4>
    <hr />
    <div
      className="alert alert-warning"
    >
      Hero 
      <b>
        1jk23gb4kj123h4gv
      </b>
       was not found
    </div>
  </div>
</div>
`;
